### ISSUE TITLE
Configure what ImGui’s vertex structure looks like?

### ISSUE THREAD

**Issue by dev0-1** [ocornut/imgui#2126]
(I think this fits here, sorry if it doesn't)
Hi,

I am building a GameEngine, my GameEngine has a canvas which is meant to draw triangles, it simply has a function that takes a vertex buffer and indice buffer, but the structure of the vertices it takes has to be the same as the one of the canvas/renderer. Now I have other libraries that use the canvas rendering functions, which work fine because it was developed for those libraries. But now I want to use ImGui as Editor GUI, and the vertex structures don’t match.
Where ImGui uses 4 bytes/chars(or one uint32) for storing colors, I use 4 floats(R, G, B, A). This makes my Engine incompatible with ImGui. Now I thought of having a way to configure the input vertex structure for the Canvas, to match whatever a library outputs.(e.g. ImGui). But this would be an incredible amount of work for me, as that means I have one vertex buffer, that holds multiple different structures.

So what I am asking here is, is it possible to alter the structure of an ImGui vertex(ImDrawVert) to match the Engine vertex structure? Preferable without modifying the ImGui code.
And, if that doesn’t exist, would it be possible for ImGui to integrate such thing?

Thanks!

Edit:
So basically what I am asking is a way to provide your own vertex structure to ImGui that it will then use. I am sure this will help a lot of people with integrating ImGui in their engines.

**Comment by ocornut** [comment]
It is possible to configure the vertex structure, you will see the possibility if you look at ImDrawVert. But it is not possible to replace the ImU32 color component of the vertice with 4 float as you are asking. 

You can either do a conversion on your side after you receive the vertices in the ImDrawData structure, either support packed color in your engine. If the later is too complicated to implement, the earlier should be a trivial for loop. See the Allegro5 example for unpacking colors on the CPU.

**Comment by dev0-1** [comment]
> 
> It is possible to configure the vertex structure, you will see the possibility if you look at ImDrawVert. But it is not possible to replace the ImU32 color component of the vertice with 4 float as you are asking. You can either do a conversion on your side after you receive the vertices in the ImDrawData structure, either support packed color in your engine. If the later is too complicated to implement, the earlier should be a trivial for loop. See the Allegro5 example for unpacking colors on the CPU.

How hard would it be to implement support for multiple structures in ImGui?

The Allegro5 example I believe loops through all the vertices and converts them to the correct format, but I am trying to avoid that because it's slow.
Currently I'm implementing a new drawTriangle function that allows you to specify what structure your vertices are. But there is one thing, I thought the ImU32 color component were actually 4 chars(R, G, B, A), is that correct? As I believe that's how they are used in the OpenGL3 example.

**Comment by ocornut** [comment]
> How hard would it be to implement support for multiple structures in ImGui?

It’s not something I would want to develop nor maintain.

> Currently I'm implementing a new drawTriangle function that allows you to specify what structure your vertices are. 

I believe this is what you should do, yes.

I don’t understand the other question. The 32 bit packed color indeed store 4 components 8 bits each. I’m sure you’ll figure it out.

**Comment by dev0-1** [comment]
> 
> 
> > How hard would it be to implement support for multiple structures in ImGui?
> 
> It’s not something I would want to develop nor maintain.
> 
> > Currently I'm implementing a new drawTriangle function that allows you to specify what structure your vertices are.
> 
> I believe this is what you should do, yes.
> 
> I don’t understand the other question. The 32 bit packed color indeed store 4 components 8 bits each. I’m sure you’ll figure it out.

But why is it one variable when it could be 4 separate?

And I understand you don't want to maintain another feature, but why not develop it, how hard can it be to have 3 different ImDrawVerts that have different data types for e.g. colors be? That would be great for a lot of people.

**Comment by ocornut** [comment]
> That would be great for a lot of people.

Unless you have talked to a lot of people, you probably mean “that would be great for me.”

I am sorry I don’t have time to argue about the cost of developing/maintaining this properly nor explain what it is more convenient to pack 4 components into 1 u32. There’s a million other things to develop and maintain for imgui. This is an easy choice to make for me, and if you are developing an engine supporting different vertex format shouldn’t be difficult, let alone the fact that you can probably use the native dx11/opengl etc renderer provided for imgui and not have to deal with the limitations of your engine.

**Comment by dev0-1** [comment]
Well, I haven't actually talked to people about it, that's true. But I noticed lot's of people with e.g. UE4, Unity, Ogre3D struggling with this because of the different format.

I understand you don't have the time. I'll just continue my adjustable vertex drawTriangles function, I can't directly use OpenGL or DirectX as ImGui is used by the Editor, and the Editor is in a different DLL/SO.

Thanks anyway.
